pyk/src/pyk/tests/test_rewrite.py

```diff
@@ -0,0 +1,160 @@
+from functools import partial
+from typing import Dict, Final, Tuple
+from unittest import TestCase
+
+from ..kast import KApply, KInner, KToken, KVariable
+from ..prelude import mlAnd, mlEquals, mlEqualsTrue, mlTop
+from ..rewrite import Subst, extract_subst
+
+a, b, c = (KApply(label) for label in ['a', 'b', 'c'])
+x, y, z = (KVariable(name) for name in ['x', 'y', 'z'])
+f, g, h = (partial(KApply.of, label) for label in ['f', 'g', 'h'])
+t = KToken('t', 'Int')
+
+
+def int_eq(term1: KInner, term2: KInner) -> KApply:
+    return KApply('_==Int_', [term1, term2])
+
+
+class SubstTest(TestCase):
+
+    def test_eq(self):
+        # Given
+        test_data = (
+            ({}, {}),
+            ({'x': x}, {}),
+            ({}, {'x': x}),
```

Comment:
Are these substitutions supposed to be considered equal?

Under composition then yes, I agree they should behave identically, but they should not be considered equal I don't think. Remember this issue where it becomes important that they are treated differently: https://github.com/runtimeverification/k/issues/2559

Reply:
I don't think that substitutions should necessarily be thought of as functions, but they can be interpreted as functions. But there are multiple acceptable interpretations of them as functions.

What I really want is a `_.union(_)` operator that fails when a substitution does not compose correctly, I think it's a better name than `_.compose(_)` for this operation.

`match(KVariable('X'), KVariable('X)).union(match(KVariable('X'), KApply('f', [KVariable['X']))` should fail. If it doesn't fail, then there is no functional interpretation of `Subst` at all, because here we wolud be giving multiple values for the same input variable `X`.

So I think we should store the substitution as `{'X', KVariable('X')}`, not as `{}`, and leave it up to the individual operations no how to interpret that. `_.union(_)` will treat it as defining the value of the substitution on `X` (and so it can't be redefined to be something else), and `_.apply(_)` will treat it as the identity mapping.

The other argument for storing the identity substitution is that it's strictly more information. It's always possible to take a substitution and remove those identities. But taking one without the identity substitutions and calculating which identities to add requires having access to the original terms that we are creating substitutions for (to find out the list of their free variables).

Reply:
> What I really want is a _.union(_) operator that fails when a substitution does not compose correctly

You mean do a union of the dictionaries but return `None` if there is a key `k` such that `s1[k] != s2[k]`? I agree that only makes sense if we allow `{'x': x}`.

> it's a better name than _.compose(_) for this operation

Compose is a different operation: given two substitutions whose interpretations as total functions are `f` and `g`, it returns a substitution whose interpretation as a total function is the function composition of `f` and `g`. The two operations coincide on substitutions such that if `f(x) != g(x)`, then `f(x) = x` or `g(x) = x`.

> So I think we should store the substitution as `{'X', KVariable('X')}`, not as `{}`,

That is a straightforward solution for enabling `union`. What we lose by doing that is a normal form w.r.t. the function interpretation, i.e. `s1 == s2` iff `s1(x) == s2(x)` for any `x`. But maybe that's not an important property after all.

Let me try to propose a solution that fits all the use cases considered here. If we can't come up with such a solution, let's fall back to losing the normal form.

Reply:
I mean, the _only_ use cases we have right now are `_.union(_)` and `_.apply(_)` (and the second is probably not needed, but `_.union(_)` as I've described it is definitely needed for any sufficiently advanced rewriting engine). I don't think we have any uses of `_.compose(_)` as it is.

Like I said, I don't give a function interpretation to substitutions, they are substitutions, not functions. There are several _natural_ interpretations of them as functions, both partial and total, and we can expose conveniences for that, but we 100% do need `_.union(_)` to operate as I described.

Not having this `_.union(_)` operation makes `Subst` unusable for anyone trying to build their own pattern matching algorithms, they'll have to do one of:

- maintain the `Dict` underneath directly, and then just wrap it in a `Subst` at the end,
- Manually track the free variables in the `pattern` term, and make sure that they fail `_.compose(_)` correctly if the returned substitution for the first argument is missing entries for free variables in that `pattern`, or
- Make sure that we always have "clean-slate" variables, which is possible _sometimes_ while making sure that we maintain user-supplied variable names, but not easy in general. Maintaining user-supplied variable names is very difficult to do in general, but so far we have been able to do so.

I have chosen to sacrifice user-defined variable names here to work around this current limitation of `Subst` class: https://github.com/runtimeverification/erc20-verification/blob/master/ksummarize/src/ksummarize/util.py#L251

I don't want to do that.

Reply:
I think I see your point now.

By the way, is this operation called `union` in the literature? If not, I'd call it `merge`.

There is a function `combine_dicts` in `pyk.rewrite`. Maybe after adding this new method, it'll be possible to no longer expose it. What do you think?

> I don't think we have any uses of _.compose(_) as it is.

Let's remove it then,

Reply:
Yeah, that's exactly what `combine_dicts` was for, we can remove it.

I think of it as the `Map` in K (the operation `_Map_`, which is only defined if the keys are distinct and is otherwise bottom). [I guess we actually call that `Map.concat`](https://github.com/runtimeverification/k/blob/master/k-distribution/include/kframework/builtin/domains.md#map-concatenation).

`merge`, `concat`, and `union` all sound fine to me

Reply:
Another simpler case of why we want to be able to know quickly "was the variable present in the original pattern" is the following `terminal` heuristic:

```
def kevm_is_terminal(cterm: CTerm) -> bool:
    config, _ = cterm
    k_subst = match(KSequence([kevm_halt(), KVariable('###CONTINUATION')]), getCell(config, 'K_CELL'))
    if k_subst and type(k_subst['###CONTINUATION']) is KVariable:
        statuscode_cell = getCell(config, 'STATUSCODE_CELL')
        if type(statuscode_cell) is KApply and statuscode_cell.arity == 0:
            return True
    return False
```

Here, we get the substitution for the `<k>` cell, and want to test not only that it worked (matching), but also that the variable `###CONTINUATION` actually had a binding in the mapping. I chose variable `###CONTINUATION` to practice defensive programming (avoid name clashes when I can), but I would also want this code to work if I happened to choose the same variable name for the continuation in the `<k>` cell as the user, without having to do `vars_apart`.

Notice that with the current implementation of `Subst`, this code would not work because if the user also chose variable name `###CONTINUATION` (or I chose a variable here that ended up clashing with a user), then the check `type(k_subst['###CONTINUATION'])` would give a "no such element" error.

Reply:
Done.